packages/lisk-transaction-pool/src/queue.ts

```diff
@@ -22,18 +32,67 @@ export class Queue {
 		this._index = {};
 	}
 
+	public dequeueUntil(condition: (transaction: Transaction) => boolean): ReadonlyArray<Transaction> {
+		const { effected, uneffected } = this._transactions.reduceRight(({effected, uneffected, conditionFailedOnce}: dequeueUntilReduceObject, transaction: Transaction) => {
+			// Add transaction to the uneffected list if the condition failed for this transaction or any previous transaction
+			if (conditionFailedOnce || !condition(transaction)) {
+				return {
+					effected,
+					uneffected: [transaction, ...uneffected],
+					conditionFailedOnce: true,
+				};
+			}
+
+			// delete the index of the transaction which passed the condition
+			delete this._index[transaction.id];
+
+			return {
+				effected: [transaction, ...effected],
+				uneffected,
+				conditionFailedOnce: false,
+			};
+		}, {
+			effected: [],
+			uneffected: [],
+			conditionFailedOnce: false
+		});
+
+		this._transactions = uneffected;
+
+		return effected;
+	}
+
 	public enqueueMany(transactions: ReadonlyArray<Transaction>): void {
+		this._transactions = [...transactions, ...this._transactions];
+
+		transactions.forEach((transaction: Transaction) => {
+			this._index[transaction.id] = transaction;
+		});
 	}
 
 	public enqueueOne(transaction: Transaction): void {
+		this._transactions = [transaction, ...this._transactions];
```

Comment:
should we check if it's in the queue or not?
I think it would be better not to trust the user here (same for enqueueMany)

Reply:
We want to ensure duplicate transactions don't exist in any of the queues. Since while adding a transaction to a queue, we can't (shouldn't) check if that particular transaction exists in some other queue, I think transaction pool should have the complete responsibility of ensuring that duplicate transaction don't exist.

Reply:
Plus, we aren't trusting the end user, we are trusting the transaction pool.